### PR TITLE
Refactor relations and fix "publish to edge" workflow

### PR DESCRIPTION
**Pull request by kelkawi-a**
This PR aims to:
- Refactor the relations (admin, ui, postgresql) into separate files to make it easier to introduce new relations in the future.
- Fix the "publish to edge" workflow. With the introduction of the scaled deployment integration tests, there should be two separate integration tests run on two different models. However, the workflow runs them both on one model which causes issues. The change introduced runs the integration tests first, before running the "publish to edge" workflow and asking it to only run one of the integration test (test_charm).

### CODE REVIEW

File: src/relations/__init__.py [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+
```
Comment (pandrey2003) [inline src/relations/__init__.py:4]: Very minor:
```suggestion

```

File: src/relations/postgresql.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            event: The event triggered when the relation changed.
+        """
+        if not self.charm._state.is_ready():
+            event.defer()
+            return
+
+        if self.charm.unit.is_leader():
+            self.charm.unit.status = WaitingStatus(f"handling {event.relation.name} change")
+            self.charm._update_db_connections(event)
+            self.charm._update(event)
```
Comment (pandrey2003) [inline src/relations/postgresql.py:53]: Very minor, for readability:
```suggestion
        if not self.charm.unit.is_leader():
            return

        self.charm.unit.status = WaitingStatus(f"handling {event.relation.name} change")
        self.charm._update_db_connections(event)
        self.charm._update(event)
```

File: src/relations/ui.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        if self.charm.unit.is_leader():
+            self._provide_server_status()
+
+    def _provide_server_status(self):
+        """Provide server status to the UI charm."""
+        charm = self.charm
+        server_status = charm.model.unit.status == ActiveStatus()
```
Comment (pandrey2003) [inline src/relations/ui.py:45]: `is_active`?

**Review by pandrey2003 — APPROVED** [review]
Looks good to me on the code level! Charming logic should be checked by another reviewer

File: src/charm.py [hunk 4]
```diff
@@ -88,22 +83,14 @@ def __init__(self, *args):
...
         self.db = DatabaseRequires(self, relation_name="db", database_name=DB_NAME)
-        self.framework.observe(self.db.on.database_created, self._on_database_changed)
-        self.framework.observe(self.db.on.endpoints_changed, self._on_database_changed)
-        self.framework.observe(self.on.db_relation_broken, self._on_database_relation_broken)
-
         self.visibility = DatabaseRequires(self, relation_name="visibility", database_name=VISIBILITY_DB_NAME)
-        self.framework.observe(self.visibility.on.database_created, self._on_database_changed)
-        self.framework.observe(self.visibility.on.endpoints_changed, self._on_database_changed)
-        self.framework.observe(self.on.visibility_relation_broken, self._on_database_relation_broken)
+        self.postgres_actions = Postgresql(self)
```
Comment (TeodorPt) [inline src/charm.py:89]: why is this called `_actions` but the others aren't?

File: tox.ini [hunk 5]
```diff
@@ -59,7 +59,7 @@ commands =
     isort --check-only --diff {[vars]src_path} {[vars]tst_path}
     black --check --diff {[vars]src_path} {[vars]tst_path}
     mypy {[vars]all_path} --ignore-missing-imports --follow-imports=skip --install-types --non-interactive
-    pylint {[vars]all_path} --disable=E0401,W1203,W0613,W0718,R0903,W1514,C0103,R0913,C0301,W0212,R0902,C0104,W0640
+    pylint {[vars]all_path} --disable=E0401,W1203,W0613,W0718,R0903,W1514,C0103,R0913,C0301,W0212,R0902,C0104,W0640,R0801
```
Comment (TeodorPt) [inline tox.ini:62]: what's the duplicate code it doesn't like?

File: src/relations/ui.py (hunk 3 above)
Comment (kelkawi-a) [reply src/relations/ui.py:45]: Will merge for now and update in the next PR.

File: src/charm.py (hunk 4 above)
Comment (kelkawi-a) [reply src/charm.py:89]: No reason in specific, will merge for now and remove ```_actions``` in the next PR.

**Pull request merged**